On tests that specify `--color=always` emit SVG file with stderr output

estebank opened the pull request:
Leverage `anstyle-svg`, as `cargo` does now, to emit `.svg` files instead of `.stderr` files for tests that explicitly enable color output. This will make reviewing changes to the graphical output of tests much more human friendly.

<img src="https://raw.githubusercontent.com/rust-lang/rust/b4bdb56f86e136ca63bf71dca3034200c6c25900/tests/ui/error-emitter/highlighting.svg">

estebank commented:
All credit to @epage for the library as well as @Muscraft for showing me https://github.com/rust-lang/cargo/pull/13461.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8118513153/job/22192885014) [(plain)](https://github.com/rust-lang/rust/commit/b65b353a457d364be9af382afe37e71d9c4eb49b/checks/22192885014/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_0fd85fc6-6d50-4b2d-ba7c-4771140e7d19
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=fancy-svg
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_0fd85fc6-6d50-4b2d-ba7c-4771140e7d19
GITHUB_REF=refs/pull/121877/merge
GITHUB_REF_NAME=121877/merge
GITHUB_REF_PROTECTED=false
---
#12 writing image sha256:8b21b0a093d654613ec1f1d06e15e62db9ecef265d399468cd00c59f9c815401 done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Fri Mar  1 23:41:22 UTC 2024
  network time: Fri, 01 Mar 2024 23:41:22 GMT
  network time: Fri, 01 Mar 2024 23:41:22 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
```

</details>

compiler-errors commented on src/tools/compiletest/src/runtest.rs:
```diff
@@ -4014,9 +4014,17 @@ impl<'test> TestCx<'test> {
     ) -> usize {
         let stderr_bits = format!("{}bit.stderr", self.config.get_pointer_width());
+        let force_color_svg = self.props.compile_flags.iter().any(|s| s.contains("--color"))
+            && !self.config.target.contains("windows");
```
Why is this necessary?

estebank replied:
It's not absolutely necessary, but I don't have access to a windows box to ensure this test ran *and* the colors for the ANSI escapes is subtly different between windows and non-windows. Can change that and do a full build on CI to get the windows svg file if we want to land it with it.

epage replied:
For annotate-snippets, @Muscraft  recently added a `testing-colors` feature flag to force the colors the same across all platforms so we don't have to deal with platform-specific colors (except for the places we need to to verify they are platform-specific)

compiler-errors replied:
I'm just worried that when ui testing on windows platforms, this will cause test failures since compiletest expects stdout files but there are svg files here.

estebank replied:
I *believe* that the filter I included was for *explicit* `windows` platform request from the test file. I will remove the platform check.

@epage is `testing-colors` included in 0.1.3?

epage replied:
`testing-colors` is a annotate-snippets feature.  Not really relevant until rustc is converted over but just giving an idea of strategies.

compiler-errors replied:
> I believe that the filter I included was for explicit windows platform request from the test file. I will remove the platform check.

For the record, the variable `self.config.target` is set for all UI tests, so I believe this will produce observable behavioral differences in compiletest, and not just for tests tagged `// only-windows`/`// ignore-windows`.

> I will remove the platform check.

Cool, that sounds like the easiest solution.

Muscraft replied:
Another solution to this might be to pull the colors out into a separate crate and share them between `rustc` and `annotate-snippets`. I have thought about doing this before, but have never talked with enough people to see if it was a good idea. It would make it easier to share the colors during the transition period and allow other crates to match colors in the long run.

compiler-errors commented:
@rustbot author

should be good to go after fixing and blessing the window test, but i'd like to take a look at the final results

estebank commented:
@bors try

bors commented:
:hourglass: Trying commit d8961f00d0aac9fcf144ddf29316ce8aa7d35a09 with merge 63ee2ef9d16777b66418d517588fd1cf0732f2cb...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d8961f00d0aac9fcf144ddf29316ce8aa7d35a09","merge_sha":"63ee2ef9d16777b66418d517588fd1cf0732f2cb"} -->

compiler-errors commented:
For the record, I think `bors try` just builds artifacts for x86_64-unknown-linux-gnu

compiler-errors reviewed (approved):
r=me when ci is green

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8125521365/job/22208226166) [(plain)](https://github.com/rust-lang/rust/commit/692f8f1118d665e58d964b9a6d26e1a2b6b5ddc8/checks/22208226166/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_cea6c629-6f05-4732-8540-a79b6d730d61
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=fancy-svg
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_cea6c629-6f05-4732-8540-a79b6d730d61
GITHUB_REF=refs/pull/121877/merge
GITHUB_REF_NAME=121877/merge
GITHUB_REF_PROTECTED=false
---
#12 writing image sha256:e11940a8e1ffc9dd0fc1e0aea857e43f64ebd7a4379ad6668f7ccb487c7275d4 done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Sat Mar  2 21:27:21 UTC 2024
  network time: Sat, 02 Mar 2024 21:27:21 GMT
  network time: Sat, 02 Mar 2024 21:27:21 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
failures:

---- [ui] tests/ui/diagnostic-flags/colored-session-opt-error.rs stdout ----
normalized stderr:
warning: -C remark requires "-C debuginfo=n" to show source locations



The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/diagnostic-flags/colored-session-opt-error/colored-session-opt-error.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args diagnostic-flags/colored-session-opt-error.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/diagnostic-flags/colored-session-opt-error.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/diagnostic-flags/colored-session-opt-error" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/diagnostic-flags/colored-session-opt-error/auxiliary" "-Cremark=foo" "--error-format=human" "--color" "always"
--- stderr -------------------------------
--- stderr -------------------------------
warning: -C remark requires "-C debuginfo=n" to show source locations



failures:
```

</details>

estebank committed:
On tests that specify --color=always emit SVG file with stderr output

Leverage `anstyle-svg`, as `cargo` does now, to emit `.svg` files
instead of `.stderr` files for tests that explicitly enable color
output. This will make reviewing changes to the graphical output of
tests much more human friendly.

estebank commented:
@bors r=compiler-errors

bors commented:
:pushpin: Commit b4bdb56f86e136ca63bf71dca3034200c6c25900 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors b4bdb56f86e136ca63bf71dca3034200c6c25900 -->
<!-- homu: {"type":"Approved","sha":"b4bdb56f86e136ca63bf71dca3034200c6c25900","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit b4bdb56f86e136ca63bf71dca3034200c6c25900 with merge a09d91b04a32def152afee94ed47933c03a84d0c...
<!-- homu: {"type":"BuildStarted","head_sha":"b4bdb56f86e136ca63bf71dca3034200c6c25900","merge_sha":"a09d91b04a32def152afee94ed47933c03a84d0c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8128412128/job/22215782493)
Approved by: compiler-errors
Pushing a09d91b04a32def152afee94ed47933c03a84d0c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8128412128/job/22215782493"},"merge_sha":"a09d91b04a32def152afee94ed47933c03a84d0c"} -->

rust-timer commented:
Finished benchmarking commit (a09d91b04a32def152afee94ed47933c03a84d0c): [comparison URL](https://perf.rust-lang.org/compare.html?start=3793e5ba23ca6a57ace8e3e267c22001cc5ef794&end=a09d91b04a32def152afee94ed47933c03a84d0c&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=3793e5ba23ca6a57ace8e3e267c22001cc5ef794&end=a09d91b04a32def152afee94ed47933c03a84d0c&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.3%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3793e5ba23ca6a57ace8e3e267c22001cc5ef794&end=a09d91b04a32def152afee94ed47933c03a84d0c&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.7%  | [2.2%, 3.2%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.5% | [-4.9%, -4.0%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=3793e5ba23ca6a57ace8e3e267c22001cc5ef794&end=a09d91b04a32def152afee94ed47933c03a84d0c&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.1% | [-1.1%, -1.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.1% | [-1.1%, -1.1%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=3793e5ba23ca6a57ace8e3e267c22001cc5ef794&end=a09d91b04a32def152afee94ed47933c03a84d0c&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 650.632s -> 650.052s (-0.09%)
**Artifact size**: 175.02 MiB -> 175.05 MiB (0.02%)
